apply.c

```diff
@@ -4755,21 +4779,33 @@ static int apply_patch(struct apply_state *state,
 		return -128;
 	offset = 0;
 	while (offset < buf.len) {
-		struct patch *patch;
 		int nr;
+		struct patch *patch;
+		struct promise_t *parse_chunk_promise;
 
 		CALLOC_ARRAY(patch, 1);
 		patch->inaccurate_eof = !!(options & APPLY_OPT_INACCURATE_EOF);
 		patch->recount =  !!(options & APPLY_OPT_RECOUNT);
-		nr = parse_chunk(state, buf.buf + offset, buf.len - offset, patch);
-		if (nr < 0) {
+
+		parse_chunk_promise = promise_init();
+		parse_chunk(state, buf.buf + offset, buf.len - offset, patch, parse_chunk_promise);
+
+		promise_assert_finished(parse_chunk_promise);
+
+		if (parse_chunk_promise->state == PROMISE_FAILURE) {
+			int nr;
+			error("\n\t%s", parse_chunk_promise->result.failure_result.message.buf);
+
+			nr = parse_chunk_promise->result.failure_result.status;
 			free_patch(patch);
 			if (nr == -128) {
 				res = -128;
 				goto end;
 			}
 			break;
 		}
```

Comment:
Please note that the original code prints _nothing_ if `nr < 0` unless it is `-128`. That's because it is not an error for a patch to have non-diff lines at the end; They are simply ignored. Therefore, you need to make sure to print the error only if we are actually erroring out, i.e.:

```diff
diff --git a/apply.c b/apply.c
index a5ad0355b5d..546ae4728bd 100644
--- a/apply.c
+++ b/apply.c
@@ -4794,11 +4794,10 @@ static int apply_patch(struct apply_state *state,
 
 		if (parse_chunk_promise->state == PROMISE_FAILURE) {
 			int nr;
-			error("\n\t%s", parse_chunk_promise->result.failure_result.message.buf);
-
 			nr = parse_chunk_promise->result.failure_result.status;
 			free_patch(patch);
 			if (nr == -128) {
+				error("\n\t%s", parse_chunk_promise->result.failure_result.message.buf);
 				res = -128;
 				goto end;
 			}
```

With that applied, t3701 passes again for me.